services/client/services.go

```diff
@@ -0,0 +1,297 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package client
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	placementproto "github.com/m3db/m3cluster/generated/proto/placement"
+	"github.com/m3db/m3cluster/kv"
+	"github.com/m3db/m3cluster/services"
+	"github.com/m3db/m3cluster/services/placement/service"
+	"github.com/m3db/m3x/watch"
+	"github.com/uber-go/tally"
+)
+
+const (
+	defaultGaugeInterval = 10 * time.Second
+)
+
+var (
+	errNotImplemented   = errors.New("not implemented")
+	errNoServiceFound   = errors.New("no service found")
+	errWatchInitTimeout = errors.New("service watch init time out")
+	errNoServiceName    = errors.New("no service specified")
+	errNoZone           = errors.New("no zone specified")
+	errNoEnv            = errors.New("no env specified")
+)
+
+// NewServices returns a client of Services
+func NewServices(opts Options) services.Services {
+	s := &client{
+		pManagers: map[string]*placementManager{},
+		opts:      opts,
+	}
+	return s
+}
+
+type client struct {
+	sync.RWMutex
+
+	opts      Options
+	pManagers map[string]*placementManager
+}
+
+func (s *client) PlacementService(sid services.ServiceID, opts services.PlacementOptions) (services.PlacementService, error) {
+	if err := validateRequest(sid); err != nil {
+		return nil, err
+	}
+
+	return service.NewPlacementService(s, sid, opts), nil
+}
+
+func (s *client) Advertise(ad services.Advertisement) error {
+	return errNotImplemented
+}
+
+func (s *client) Unadvertise(sid services.ServiceID, id string) error {
+	return errNotImplemented
+}
+
+func (s *client) Query(sid services.ServiceID, opts services.QueryOptions) (services.Service, error) {
+	if err := validateRequest(sid); err != nil {
+		return nil, err
+	}
+
+	v, err := s.placement(sid)
+	if err != nil {
+		return nil, err
+	}
+
+	service, err := getServiceFromValue(v, sid)
+	if err != nil {
+		return nil, err
+	}
+	return service, nil
+}
+
+func (s *client) Watch(sid services.ServiceID, opts services.QueryOptions) (xwatch.Watch, error) {
+	if err := validateRequest(sid); err != nil {
+		return nil, err
+	}
+
+	s.opts.InstrumentsOptions().Logger().Infof(
+		"adding a watch for service: %s env: %s zone: %s includeUnhealthy: %v",
```

Comment:
This barely logs, most times only log once when a service starts up, seems to be helpful to me